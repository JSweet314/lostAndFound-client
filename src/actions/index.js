export const captureUser = user => ({
  type: 'CAPTURE_USER',
  user
});

export const logOutUser = () => ({
  type: 'LOG_OUT_USER'
});

export const submitNewUser = user => ({
  type: 'SUBMIT_NEW_USER',
  user
});

export const captureErrorMessage = message => ({
  type: 'CAPTURE_ERROR_MESSAGE',
  message
});

export const clearErrorMessage = () => ({
  type: 'CLEAR_ERROR_MESSAGE'
});