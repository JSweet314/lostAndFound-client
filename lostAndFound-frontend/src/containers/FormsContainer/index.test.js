import React from 'react';
import { shallow } from 'enzyme';
import { FormsContainer, mapDispatchToProps, mapStateToProps } from './index';
import LocalStorage from '../../__mocks__/localStorageMock';
import * as actions from '../../actions';

window.localStorage = new LocalStorage();

describe('FormsContainer', () => {
  let wrapper;
  const mockUser = { username: 'tom', password: 'abc', email: 'tom@gmail.com' };
  const mockEvent = { preventDefault: jest.fn() };
  const mockChangeEvent = { target: { name: 'name', value: 'tom' } };
  const mockLoginMatch = { params: { id: 'signup' } };
  const mockSubmitNewUser = jest.fn();
  const mockSignInUser = jest.fn();
  const mockCaptureUser = jest.fn();
  beforeEach(() => {
    wrapper = shallow(
      <FormsContainer
        user={{}}
        loggedIn={false}
        captureUser={mockCaptureUser}
        submitNewUser={mockSubmitNewUser}
        signInUser={mockSignInUser}
        match={mockLoginMatch}
      />
    );
  });

  it('should match a snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should save changing form input in state', () => {
    wrapper.instance().handleOnChange(mockChangeEvent);
    expect(wrapper.state('name')).toEqual('tom');
  });

  it('should call submitNewUser when a new user submits a form', () => {
    wrapper.setState(mockUser);
    wrapper.instance().handleOnSubmit(mockEvent);
    expect(mockSubmitNewUser).toHaveBeenCalledWith(mockUser);
  });

  it('should reset the state when a form is submitted', () => {
    wrapper.setState(mockUser);
    wrapper.instance().handleOnSubmit(mockEvent);
    expect(wrapper.state()).toEqual(
      { username: '', email: '', password: '', showPassword: false }
    );
  });

  it('should store a logged in user in localStorage after signin', () => {
    window.localStorage.setItem = jest.fn();
    wrapper = shallow(
      <FormsContainer
        user={{loggedIn: true}}
        loggedIn={true}
        captureUser={mockCaptureUser}
        submitNewUser={mockSubmitNewUser}
        signInUser={mockSignInUser}
        match={mockLoginMatch}
      />
    );
    wrapper.instance().componentDidUpdate({loggedIn: false});
    const expected = [
      'LFUser',
      JSON.stringify({ loggedIn: true })
    ];
    expect(window.localStorage.setItem).toHaveBeenCalledWith(...expected);
  });

  it('should be able to toggle password visibility', () => {
    wrapper.instance().togglePasswordVisibility(mockEvent);
    expect(wrapper.state('showPassword')).toBe(true);
  });

  describe('mapDispatchToProps', () => {
    const mockDispatch = jest.fn();
    it('should call dispatch with action creator captureUser', () => {
      const mapped = mapDispatchToProps(mockDispatch);
      mapped.captureUser(mockUser);
      expect(mockDispatch).toHaveBeenCalledWith(actions.captureUser(mockUser));
    });

    it('should call dispatch with action creator submitNewUser', () => {
      const mapped = mapDispatchToProps(mockDispatch);
      mapped.submitNewUser(mockUser);
      expect(mockDispatch)
        .toHaveBeenCalledWith(actions.submitNewUser(mockUser));
    });

    it('should call dispatch with action creator signInUser', () => {
      const mapped = mapDispatchToProps(mockDispatch);
      mapped.signInUser(mockUser);
      expect(mockDispatch).toHaveBeenCalledWith(actions.signInUser(mockUser));
    });
  });

  describe('mapStateToProps', () => {
    const mockState = {user: {loggedIn: false}};
    
    it('should return an object of data from the store', () => {
      const expected = { 
        "loggedIn": false, 
        "user": { 
          "loggedIn": false 
        } 
      };
      expect(mapStateToProps(mockState)).toEqual(expected);
    });
  });
});