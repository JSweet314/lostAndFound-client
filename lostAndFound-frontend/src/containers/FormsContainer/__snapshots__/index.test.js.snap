// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormsContainer should match a snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormsContainer
    captureUser={[Function]}
    loggedIn={false}
    match={
        Object {
            "params": Object {
              "id": "signup",
            },
          }
    }
    signInUser={[Function]}
    submitNewUser={[Function]}
    user={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "email": "",
      "handleOnChange": [Function],
      "handleOnSubmit": [Function],
      "password": "",
      "routeId": "signup",
      "showPassword": false,
      "togglePasswordVisibility": [Function],
      "username": "",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "email": "",
        "handleOnChange": [Function],
        "handleOnSubmit": [Function],
        "password": "",
        "routeId": "signup",
        "showPassword": false,
        "togglePasswordVisibility": [Function],
        "username": "",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
